plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2536,9 +2668,184 @@ void producer_plugin::log_failed_transaction(const transaction_id_type& trx_id,
+         // Do not schedule new transactions
+         break;
+      }
+   }
+
+   // If all tasks are finished, do not wait until end of read window; switch to write window now.
+   if ( --_ro_num_active_trx_exec_tasks == 0 ) {
+      // do switching on app thread
+      app().executor().post( priority::high, exec_queue::read_only_trx_safe, [self=this]() {
+         self->_ro_trx_exec_tasks_fut.clear();
+         self->switch_to_write_window();
+      } );
+   }
+
+   return true;
+}
+
+// Called from a read only trx thread. Run in parallel with app and other read only trx threads
+// Return whether the trx needs to be retried in next read window
+bool producer_plugin_impl::process_read_only_transaction(const packed_transaction_ptr& trx, const next_function<transaction_trace_ptr>& next, const fc::time_point& read_window_start_time) {
+   chain::controller& chain = chain_plug->chain();
+   auto future = transaction_metadata::start_recover_keys( trx, _thread_pool.get_executor(),
+                                                           chain.get_chain_id(), fc::microseconds::maximum(),
+                                                           transaction_metadata::trx_type::read_only,
+                                                           chain.configured_subjective_signature_length_limit() );
+   auto exception_handler = [&next](fc::exception_ptr ex) {
+      next( std::move(ex) );
+      return false;
+   };
+   future.wait();
+   try {
+      auto trx_metadata = future.get();
+      return push_read_only_transaction( trx_metadata, next, read_window_start_time );
+   } CATCH_AND_CALL(exception_handler);
+   return false;
+}
+
+// Called on read_only_trx execution thread; multi-threaded safe
+// Return whether the trx needs to be retried in next read window
+bool producer_plugin_impl::push_read_only_transaction(
+            const transaction_metadata_ptr& trx,
+            next_function<transaction_trace_ptr> next,
+            const fc::time_point& read_window_start_time) {
+   auto retry = false;
+   chain::controller& chain = chain_plug->chain();
+
+   if( !chain.is_building_block() ) {
+      // try next round
+      return true;
+   }
+
+   try {
+      const auto block_deadline = calculate_block_deadline( chain.pending_block_time() );
+      auto start = fc::time_point::now();
+      auto time_used_in_read_window_us = ( start - read_window_start_time );
+      if ( time_used_in_read_window_us >= _ro_read_window_time_us ) {
+         // already passed read-window deadline, try next round
+         return true;
+      }
+
+      auto remained_time_in_read_window_us = _ro_read_window_time_us - time_used_in_read_window_us;
+      // set the trx to finish by the end of read-window. limit read window overrun
+      auto available_trx_time_us = std::min( remained_time_in_read_window_us, _ro_max_trx_time_us );
+
+      auto trace = chain.push_transaction( trx, block_deadline, available_trx_time_us, 0, false, 0 );
```

Comment:
This is setting an "objective" trx time limit via `available_trx_time_us` which will cause the trx to fail if it does not execute in that time. Instead you want to provide the min of `remained_time_in_read_window_us` and `block_deadline`, call it `window_deadline`,  as the `block_deadline` and `_ro_max_trx_time_us` as the trx deadline.

Reply:
Thanks. Your comment clears my confusion. Will change now.